src/scores/continuous/threshold_weighted_impl.py

```diff
@@ -0,0 +1,657 @@
+"""
+Implementation of some basic threshold weighted scoring functions.
+See Taggart (2022) https://doi.org/10.1002/qj.4206
+"""
+
+import functools
+from typing import Callable, Literal, Optional, Tuple, Union
+
+import numpy as np
+import xarray as xr
+
+from scores.continuous.consistent_impl import (
+    check_alpha,
+    check_huber_param,
+    consistent_expectile_score,
+    consistent_huber_score,
+    consistent_quantile_score,
+)
+from scores.typing import FlexibleDimensionTypes
+
+HUBER_FUNCS = ["huber_loss", "scaled_huber_loss"]
+ALPHA_FUNCS = ["quantile_score", "expectile_score"]
+OTHER_FUNCS = ["squared_error", "absolute_error"]
+SCORING_FUNCS = OTHER_FUNCS + ALPHA_FUNCS + HUBER_FUNCS
+AuxFuncType = Callable[[xr.DataArray], xr.DataArray]
+EndpointType = Union[int, float, xr.DataArray]
+
+
+def threshold_weighted_score(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    scoring_func: Literal[
```

Comment:
Is the best as a string, as a function pointer, or as multiple  functions (e.g. threshold_weighted_squared_error, threshold_weighted_absolute_error), or as a base class with classes. e.g. class TW_SquaredError(ThresholdWeightedScore).  I'm not saying the current approach is wrong, but whenever I see strings being passed as selectors I think it's worth asking the question. Or, potentially, scores.continuous.tw.squared_error, scores.continuous.tw.huber_loss etc.

Reply:
A couple of reasons for a function rather than a class:
1. It follows the pattern of all the metrics in scores (the exception is the contingency table class which allows multiple metrics to be calculated from a contingency table).
2. Backwards compatibility with Jive.

I have some convenience functions towards the end of this file that align with your last suggestion. More could be added in the future.

Reply:
For clarity, the downsides (mainly to API users) of using strings in this fashion are as follows:

- Unlike a function name or a function pointer, code completion and static analysis doesn't understand what's in the string
- The API is not as self-documenting / discoverable in terms of having clear API endpoints for each supported threshold-weightable function
- Things like typos in the strings aren't going to be picked up with a clear warning message as would be the case  with a mistyped function name or a mistyped function argument
- Inconsistent with other approaches in the codebase - e.g. the consist_impl files has e.g. consistent_quantile_score, consistent_huber_score rather than a general consistenct_sore(score_type='huber')
- Ability to separately provide docstrings for each threshold-weighted score which describe that score specifically
- Allows test functions to be more specific to just one score type
- It's not possible to pass in a novel score which passes custom values to consistent_expectile_score in a new way. While users could call consistent_expectile_score directly, they would also have to then take on the work of checking arguments and extracting g, phi and phi_prime, resulting in potential inconsistencies.

On option would be to make threshold_weighted_score a private method (i.e. not present in the public API) and only make the convenience functions part of the public API. 

The actual functionality in threshold_weighted_score appears to be to check the arguments, and extract g, phi and phi_prime. The scoring func could be a callable rather than a string, perhaps defined as functions in the module. If threshold_weighted_score is mainly a worker engine for the convenience functions, then the arguments against using a string selector are even stronger. 

If any of those suggestions seems reasonable, I'm happy to refactor the code and put that up for you to consider.

I also have just had a lot of problems with string-passing code generally, which doesn't directly apply here but has left me generally with the view that using strings to control logic flows of what's inside a function is likely to make the code harder to understand.

I can be convinced differently, but I think my perspective would be to make the convenience functions the primary recommended API for users, put them in a scores.continuous.tw module, probably also prefix the methods with e.e.g tw_absolute_error, tw_quantile_score so it's still clear if people are just using the score name, keep the general threshold_weights_score method as an underlying engine, but pass in callables rather than strings.

Reply:
For completeness, I'm totally fine with not using classes here, I'm just talking about making the most elegant API for users which isn't too brittle / error-prone.

Reply:
My plan is:
- Get rid of `threshold_weighted_score`. There will need to be some updates in Jive.
- Have convenience functions (6 of them) instead. `_check_tws_args` and `_auxiliary_funcs` is called in each one.
- Move the second two checks to the relevant functions.

This will remove strings from being passed around in all places.

I'll push up a commit where this has been done for tw_squared_error with tests passing, but will continue on with refactoring the rest of the scores until I hear otherwise.